system-addon/content-src/components/LinkMenu/LinkMenu.jsx

```diff
@@ -1,52 +1,98 @@
 const React = require("react");
 const {injectIntl} = require("react-intl");
 const ContextMenu = require("content-src/components/ContextMenu/ContextMenu");
+const {actionTypes, actionCreators} = require("common/Actions.jsm");
 
 class LinkMenu extends React.Component {
-  constructor(props) {
-    super(props);
-    this.getOptions = this.getOptions.bind(this);
+  getBookmarkStatus(site) {
+    return (site.bookmarkGuid ? {
+      id: "menu_action_remove_bookmark",
+      icon: "bookmark-remove",
+      action: actionTypes.DELETE_BOOKMARK_BY_ID,
+      data: site.bookmarkGuid
+    } : {
+      id: "menu_action_bookmark",
+      icon: "bookmark",
+      action: actionTypes.BOOKMARK_URL,
+      data: site.url
+    });
+  }
+  getDefaultContextMenu(dispatch, site) {
+    return [{
+      id: "menu_action_open_new_window",
+      icon: "new-window",
+      onClick: () => {
+        dispatch(actionCreators.SendToMain({type: actionTypes.OPEN_NEW_WINDOW, data: {url: site.url}}));
+      }
+    },
+    {
+      id: "menu_action_open_private_window",
+      icon: "new-window-private",
+      onClick: () => {
+        dispatch(actionCreators.SendToMain({type: actionTypes.OPEN_PRIVATE_WINDOW, data: {url: site.url, isPrivate: true}}));
+      }
+    }];
   }
   getOptions() {
-    return [
+    const {dispatch, site} = this.props;
+    const defaultOptions = this.getDefaultContextMenu(dispatch, site);
+
+    // default top sites have a limited set of context menu options
+    if (site.isDefault) {
+      return this.getFirstAndLast(defaultOptions).map(option => {
+        // Convert message ids to localized labels
```

Comment:
Instead of duplicating the `intl` code or needing the `getFirstAndLast` helper function, this code will probably flow better without the short circuit:
```js
getOptions() {
  let options = getDefaultContextMenu();
  // Add more options for regular sites
  if (!site.isDefault) {
    options = [
      { bookmarkStuff },
      { separator },
      ...options, // inserts all default options here with spread operator
      { separator },
      …
    ];
  }
  // convert ids to labels and hook up onClick action
  options.map(option => …);
  // annotate the first and last options for a11y
  options[0].first = true;
  …
}
```